custom_metrics/sass.js

```diff
@@ -0,0 +1,309 @@
+			calls: scss.match(RegExp("\\b" + name + "\\(", "gi"))?.length
+		};
+	});
+
+	// TODO Measure usage of Sass functions
+	ret.functionCalls = {};
+	let functionCallRegex = new RegExp(`(?<![-.\\w])(${SassFunctions.join("|").replace(".", "\\.")})\\(`, "gi");
+	scss.replace(functionCallRegex, ($0, name) => {
+		ret.functionCalls[name] = (ret.functionCalls[name] || 0) + 1;
+	});
+	ret.functionCalls = sortObject(ret.functionCalls);
+
+	// Conditionals
+	ret.ifs = [];
+	scss.replace(/@(?:(else )?if\s+?(.+?)|@else)(?=\s*\{)/g, ($0, elseIf, test) => {
+		if ($0 === "@else") {
+			// Else for previous condition
+			let obj = ret.ifs.pop();
+			console.log(ret.ifs, obj);
+			obj.hasElse = true;
+			ret.ifs.push(obj);
+		}
+		else if (elseIf) {
+			// Else if for previous condition
+			let obj = ret.ifs.pop();
+			obj.elseIfs = obj.elseIfs || [];
+			obj.elseIfs.push(test);
+			ret.ifs.push(obj);
+		}
+		else {
+			// New conditional
+			ret.ifs.push({test});
+		}
+	});
+
+	// @each, @for, @while Loops
+	ret.eaches = {};
+	scss.replace(/@each\s+(.+ in .+)(?=\s*\{)/g, ($0, args) => {
+		ret.eaches[args] = (ret.eaches[args] || 0) + 1;
+	});
+
+	ret.fors = {};
+	// console.log("@for", scss);
+	scss.replace(/@for\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.fors[args] = (ret.fors[args] || 0) + 1;
+	});
+
+	ret.whiles = {};
+	scss.replace(/@while\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.whiles[args] = (ret.whiles[args] || 0) + 1;
+	});
+
+	// @extend
+	ret.extends = {};
+	scss.replace(/@extend\s*(.+)(?=\s*(?:[;}]|$))/g, ($0, selector) => {
+		ret.extends[selector] = (ret.extends[selector] || 0) + 1;
+	});
+
+	// @error
+	ret.errors = scss.match(/@error (.+)/g)?.map(e => e.slice(7));
+
+	// CSS variables that are set with Sass variables
+	ret.variablesCombined = {};
+	scss.replace(/(?<=^|\s)--([\w-]+):\s*(.*#\{.*\$.+\}.*)\s*$/gm, ($0, name, value) => {
+		ret.variablesCombined["--" + name] = value;
+	});
+
+	// Heuristic for nesting &
+	ret.nested = {
+		"descendant": scss.match(/&\s+(?=[\w[.:#])/g)?.length,
+		">": scss.match(/&\s*(?=>)/g)?.length,
+		"+": scss.match(/&\s*(?=\+)/g)?.length,
+		"~": scss.match(/&\s*(?=~)/g)?.length,
+		"lone": scss.match(/&\s*(?=[,{]|$})/g)?.length,
+		"pseudo-element": scss.match(/&::/g)?.length,
+		"pseudo-class": scss.match(/&:(?!:)/g)?.length,
+		"id": scss.match(/&#\w+/g)?.length,
+		"class": scss.match(/&\.\w+/g)?.length,
+		"attr": scss.match(/&\[\w+/g)?.length,
+	};
+}
+
+let results = (async () => {
+
+const results = {};
+const $$ = s => [...document.querySelectorAll(s)];
+const sourcemapRegex = /\/\*[#@] sourceMappingURL=(.+?) \*\//;
+
+let stylesheets;
+
+if (typeof $WPT_BODIES !== "undefined") {
+	// in WPT, we already have the bodies
+	stylesheets = $WPT_BODIES.filter(request => request.type == "Stylesheet")
+		.map(file => {
+			return {url: file.url, body: file.response_body};
+		});
+}
+else {
+	// Outside WPT, use fetch()
+	stylesheets = await Promise.all($$("link[rel=stylesheet]").map(async link => {
```

Comment:
This code will always run in WPT. Could we remove this `else`? Alternatively, is it feasible to get what we need using qSA and fetch?

Reply:
Sure, if code size matters. I have it there because this way it's easy to test the code with a simple `import()` in the console, without the whole wait time of WPT. But no problem removing it!

Reply:
That'd be great, thanks.